<HTML><HEAD><META HTTP-EQUIV="Content-Type" CONTENT="text/html; charset=iso-8859-1">
<!--SafClassName="docSection1Title"--><!--SafTocEntry="C.7 Chapter 7"-->
<LINK REL="stylesheet" href="FILES/sbolM.css">
</HEAD>
<BODY link="#354278" vlink="#000066" alink="#000080" topmargin="0">
<table width="100%" border="0" cellspacing="0" cellpadding="0" bgcolor="#CCCCCC">
<td><font color="black" size=1>I l<font color="#FF0000">@</font>ve RuBoard</td>
<td valign="top" class="v2" align="right"><a href="lpython_snode137.html"><img src="FILES/previous.gif" width="62" height="15" border="0" align="absmiddle" alt="Previous Section"></a> <a href="lpython_snode139.html"><img src="FILES/next.gif" width="41" height="15" border="0" align="absmiddle" alt="Next Section"></a></td></table>
<FONT>
<h3>C.7
<a href="lpython_cnode77.html">Chapter 7</A></H3>


<ol tyPE="1">
<LI><p><i>try/except</i>. Our
version of the <tT CLAss="monofont">oops</tt> function follows. As for the
noncoding questions, changing <tt class="monofont">oops</tt> to raise
<tt class="monofont">KeyError</Tt> instead of <Tt clAsS="monofont">IndexError</Tt>
means that the exception won't be caught by our try handler (it
"percolates" to the top level and triggers Python's
default error message). The names <TT cLaSS="monofont">KeyError</TT> and
<tt clASS="monofont">IndexError</Tt> come from the outermost built-in names
scope. If you don't believe us, import<tt cLASS="monofont"> </tt>_
_<tt CLASs="monofont">builtin</tt>__ and pass it as an argument to the
<tt class="monofont">dir</tt> function to see for yourself.</p>



<pre clasS="monofont">%<b> cat oops.py</B>

def oops():
    raise IndexError

def doomed():
    try:
        oops()
    except IndexError:
        print 'caught an index error!'
    else:
        print 'no error caught...'

if __name__ == '__main__': doomed()

%<b> python oops.py</b>
caught an index error!</pRe></LI>
<lI><P><i>Exception lists</I>.
Here's the way we extended this module for an exception of our
own:</p>




<PRE Class="monofont">%<B> cat oops.py</B>
MyError = 'hello'

def oops():
    raise MyError, 'world'

def doomed():
    try:
        oops()
    except IndexError:
        print 'caught an index error!'
    except MyError, data:
        print 'caught error:', MyError, data
    else:
        print 'no error caught...'

if __name__ == '__main__':
    doomed()

%<B> python oops.py</B>
caught error: hello world</pre></lI>
<LI><P><i>Error handling</i>.
Finally, here's one way to solve this one; we decided to do our
tests in a file, rather than interactively, but the results are about
the same.</p>




<pRE CLass="monofont">%<b> cat safe2.py</b>
import sys, traceback

def safe(entry, *args):
    try:
        apply(entry, args)                 # catch everything else
    except:
        traceback.print_exc()
        print 'Got', sys.exc_type, sys.exc_value

import oops
safe(oops.oops)

%<b> python safe2.py</b>
Traceback (innermost last):
  File "safe2.py", line 5, in safe
    apply(entry, args)                     # catch everything else
  File "oops.py", line 4, in oops
    raise MyError, 'world'
hello: world
Got hello world</pre></li>
</ol>

</font>
 
<table width="100%" border="0" cellspacing="0" cellpadding="0" bgcolor="#CCCCCC">
<td><font color="black" size=1>I l<font color="#FF0000">@</font>ve RuBoard</td>
<td valign="top" class="v2" align="right"><a href="lpython_snode137.html"><img src="FILES/previous.gif" width="62" height="15" border="0" align="absmiddle" alt="Previous Section"></a> <a href="lpython_snode139.html"><img src="FILES/next.gif" width="41" height="15" border="0" align="absmiddle" alt="Next Section"></a></td></table>
</BODY></HTML>